Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -153,567 +131,429 @@ final class ConnectionPool {
-
-        /// The key associated with this provider
-        private let key: ConnectionPool.Key
-
-        /// The `State` of this provider
-        ///
-        /// This property holds data structures representing the current state of the provider
-        /// - Warning: This type isn't thread safe and should be accessed with proper
-        /// synchronization (see the `stateLock` property)
-        fileprivate var state: State
-
-        /// The lock used to access and modify the `state` property
-        ///
-        /// - Warning: This lock should always be acquired *after* `ConnectionPool`s `connectionProvidersLock` if used in combination with it.
-        fileprivate let stateLock = Lock()
-
-        /// The maximum number of concurrent connections to a given (host, scheme, port)
-        private let maximumConcurrentConnections: Int = 8
-
-        /// The pool this provider belongs to
-        private let parentPool: ConnectionPool
-
-        /// Creates a new `HTTP1ConnectionProvider`
-        ///
-        /// - parameters:
-        ///     - key: The `Key` (host, scheme, port) this provider is associated to
-        ///     - configuration: The client configuration used globally by all requests
-        ///     - initialConnection: The initial connection the pool initializes this provider with
-        ///     - parentPool: The pool this provider belongs to
-        init(key: ConnectionPool.Key, eventLoop: EventLoop, configuration: HTTPClient.Configuration, parentPool: ConnectionPool) {
-            self.eventLoop = eventLoop
-            self.configuration = configuration
-            self.key = key
-            self.parentPool = parentPool
-            self.state = State(eventLoop: eventLoop, parentPool: parentPool, key: key)
+    /// Sets idle timeout handler and channel inactivity listener.
+    func setIdleTimeout(timeout: TimeAmount?) {
+        _ = self.channel.pipeline.addHandler(IdleStateHandler(writeTimeout: timeout)).flatMap { _ in
+            self.channel.pipeline.addHandler(IdlePoolConnectionHandler(connection: self))
         }
+    }
 
-        deinit {
-            assert(self.state.activity == .closed, "Non closed on deinit")
-            assert(self.state.availableConnections.isEmpty, "Available connections should be empty before deinit")
-            assert(self.state.leased == 0, "All leased connections should have been returned before deinit")
-            assert(self.state.waiters.count == 0, "Waiters on deinit: \(self.state.waiters)")
+    /// Removes idle timeout handler and channel inactivity listener
+    func cancelIdleTimeout() -> EventLoopFuture<Void> {
+        return self.removeHandler(IdleStateHandler.self).flatMap { _ in
+            self.removeHandler(IdlePoolConnectionHandler.self)
         }
+    }
+}
 
-        var description: String {
-            return "HTTP1ConnectionProvider { key: \(self.key), state: \(self.state) }"
-        }
+extension Connection: Hashable {
```

Comment:
Minor nit, but I'd rather not conform something to `Hashable`/`Equatable` if its definition of hash value and equality is identity.

Reply:
Why? I what would you recommend instead?

Reply:
Users explicitly testing for identity and explicitly using `ObjectIdentifier` as a hash key.

The main reason I object is that Swift pretty pervasively separates the idea of equality and identity. Indeed, the [`Equatable` docs call this out directly](https://developer.apple.com/documentation/swift/equatable#2847781). In particular, being equatable implies that two objects are substitutable. If two objects are only equal if they are literally the same object, we have a way to express that already and do not need to bring equatability into the mix.

Reply:
I don't think there is a good way to use `ObjectIdentifier` is a `Set`, while I can add and test for `contains` and remove, since `ObjectIdentifier` is not generic, I cannot say `leased.map { $0.close() }`

Reply:
Do you need a `Set`? Why not just `Array`?

Reply:
Main usage - is testing if given connection is in `leased` set

Reply:
Ok, so let's step back. We have three options:

1. Make `Connection` `Hashable`.
2. Put `ObjectIdentifier` into the `leased` `Set`.
3. Change the `leased` structure from `Set` to `Array`.

My inclination is that either (2) or (3) is preferable to (1).

(2) is acceptable if you never need to iterate the `leased` `Set` to grab `Connection`s from it. If you do, then `Array` works just fine. In general a linear `Array` search is _very_ fast, and likely to be faster than hashing for small numbers of elements. The only reason it'd be a concern is if you are having to add-and-remove elements a lot, and even there if `leased` is small then I think it's really not a big problem.

Reply:
@artemredkin you can always make a wrapper struct that implements identity hashable/equals

Reply:
(an ad-hoc one that you just use here)

Reply:
or use a dictionary `[ObjectIdentifier: Connection]`.

Reply:
@Lukasa and 4) (which I think may be best):

```
fileprivate struct LeasedConnectionKey: Hashable {
    private let connection: Connection

   func ==(...) {}
   func hash...
}
```

Reply:
Sure, works for me.

Reply:
I do add and remove connections constantly, its added on every new http call and removed when call is finished. This is a result from a simple benchmark (adds, removes, searches) on a small (8 element) array/set:
```
name   time       std         iterations  
----------------------------------------
array  3208.0 ns  ±  29.53 %  425528      
set    1677.0 ns  ±  52.07 %  827306 
```
In this case set is almost twice as fast...

I think I'll go with 4

Reply:
Yeah, the problem with Array will be the removals of elements that are not the last...

Reply:
done